---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# ISAnalytics <a href='https://bioconductor.org/packages/3.12/bioc/html/ISAnalytics.html'><img src='man/figures/logo.png' align="right" height="250" /></a>

<!-- badges: start -->
[![Travis build status](https://travis-ci.com/calabrialab/isanalytics.svg?branch=master)](https://travis-ci.com/calabrialab/isanalytics)
[![codecov](https://codecov.io/gh/calabrialab/ISAnalytics/branch/master/graph/badge.svg)](https://codecov.io/gh/calabrialab/ISAnalytics)
[![R build status - bioc](https://github.com/calabrialab/isanalytics/workflows/R-CMD-check-bioc/badge.svg)](https://github.com/calabrialab/isanalytics/actions)
[![Lifecycle: maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![BioC status](http://www.bioconductor.org/shields/build/release/bioc/ISAnalytics.svg)](https://bioconductor.org/checkResults/release/bioc-LATEST/ISAnalytics)
<!-- badges: end -->

ISAnalytics is an R package developed to analyze gene therapy vector insertion sites data identified from genomics next generation sequencing reads for clonal tracking studies.

In gene therapy, stem cells are modified using viral vectors to deliver the therapeutic transgene and replace functional properties since the genetic modification is stable and inherited in all cell progeny. The retrieval and mapping of the sequences flanking the virus-host DNA junctions allows the identification of insertion sites (IS), essential for monitoring the evolution of genetically modified cells in vivo. A comprehensive toolkit for the analysis of IS is required to foster clonal trackign studies and supporting the assessment of safety and long term efficacy in vivo. This package is aimed at (1) supporting automation of IS workflow, (2) performing base and advance analysis for IS tracking (clonal abundance, clonal expansions and statistics for insertional mutagenesis, etc.), (3) providing basic biology insights of transduced stem cells in vivo.

# Visit the package website

You can visit the package website to view documentation, vignettes and 
more.

* For the release version: [ISAnalytics Website release](https://calabrialab.github.io/ISAnalytics/)
* For the devel version: [ISAnalytics Website dev](https://calabrialab.github.io/ISAnalytics/dev)

# Installation

## Installation from bioconductor

RELEASE version:
```{r eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("ISAnalytics")
```

DEVEL version:
```{r eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

# The following initializes usage of Bioc devel
BiocManager::install(version='devel')

BiocManager::install("ISAnalytics")
```

## Installation from GitHub

RELEASE:
```{r eval=FALSE}
if (!require(devtools)) {
  install.packages("devtools")
}
devtools::install_github("calabrialab/ISAnalytics",
                         ref = "RELEASE_3_12",
                         dependencies = TRUE,
                         build_vignettes = TRUE)

## Safer option for vignette building issue
devtools::install_github("calabrialab/ISAnalytics",
                         ref = "RELEASE_3_12")
```

DEVEL:
```{r eval=FALSE}
if (!require(devtools)) {
  install.packages("devtools")
}
devtools::install_github("calabrialab/ISAnalytics",
                         ref = "master",
                         dependencies = TRUE,
                         build_vignettes = TRUE)

## Safer option for vignette building issue
devtools::install_github("calabrialab/ISAnalytics",
                         ref = "master")
```

# Current functionality

* Import integration matrices from files: more info with `vignette("How to use import functions", package = "ISAnalytics")`
* Collision removal: more info with `vignette("Collision removal functionality", package = "ISAnalytics")`
* Aggregation: more info with `vignette("Working with aggregate functions", package = "ISAnalytics")`
* Re-calibration functions: `compute_near_integrations`
* Analysis functions: `compute_abundance`, `comparison_matrix` `separate_quant_matrices`, others
* Plotting functions: `CIS_volcano_plot`
* Utility functions

# NEWS

```{r child="NEWS.md"}

```


# TO DO in future updates  

- [ ] Add further plot wrappers
- [ ] New functionality: raw reads flagging/filtering
- [ ] New functionality: HSCs population size estimate
- [ ] New functionality: genomic annotation
- [ ] Obtain a system to visualize interactive reports based on shiny (possibly
replacing self-contained HTML reports)


# Getting help

For help please contact the maintainer of the package or open an issue on GitHub.


